compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -2727,7 +2727,7 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
     else typd(imp.expr, AnySelectionProto)
 
   def typedImport(imp: untpd.Import)(using Context): Tree =
-    val sym = retrieveSym(imp)
+    val sym = imp.removeAttachment(SymOfTree).getOrElse(NoSymbol)
     val expr1 = typedImportQualifier(imp, typedExpr(_, _)(using ctx.withOwner(sym)))
```

Comment:
This could typecheck an expression in a context which has `NoSymbol` as owner, which will cause havoc if `imp` has any local symbols. 

I think we need to dig deeper. Either accept the cyclic error but avoid `<import>` in the error message or somehow heal the situation. But I disagree that a `NotFound` is necessarily better. I think it could be more puzzling than the cyclic error. 

Generally, I think trying `export` for package management is doomed. Even if it worked, I would find it dubious since it would just add another layer of complexity.

Reply:
> Generally, I think trying export for package management is doomed

That's a pity, because it seems that people naturally tend to use that way (as in https://contributors.scala-lang.org/t/why-the-current-export-is-problematic-for-import/6147), and because cyclic errors involving exports might not mention export/import at all, like in https://github.com/lampepfl/dotty/issues/17076, where the problem suddenly started manifesting itself long after the user had started using export, making it extremely hard to debug.

Reply:
Yes, I think the best thing to do now is warn people that that's an anti-pattern.

Reply:
Thanks.
I will look at that a bit more.

Having [an explanation](https://github.com/lampepfl/dotty/pull/17304#issuecomment-1522264780) from @smarter I'm wondering if we can also try to trigger package completion (which includes package export) in advance before visiting first first import. In theory it should change the completion order and at least fix #17201

Reply:
I believe any change in the order would cause other cycles. This whole thing (trying to avoid cycles) is very, very finely balanced. And it is also what it is, by now. Any change in the order would likely reject existing code, and therefore is infeasible at this point. 

One _can_ delay evaluation at some point if one can arrange things that certain info is not needed at this point and can be accessed later. For instance, a recent change of mine moved some checks to PostTyper, since they needed info that caused cycles in Typer. But I don't think that's applicable for exports.

The other room for improvement is in better diagnostics if there is a CyclicReference. I believe there is still a lot to be done, but it is not easy.

Reply:
> > This could typecheck an expression in a context which has NoSymbol as owner, which will cause havoc if imp has any local symbols.

I don't think this is a behavior change from what happened before, `retrieveSym` can also return NoSymbol, but import qualifiers don't define new symbols so I'm not sure that `withOwner` does anything anyway.

Reply:
> This could typecheck an expression in a context which has `NoSymbol` as owner, which will cause havoc if `imp` has any local symbols.
> 
> I think we need to dig deeper. Either accept the cyclic error but avoid `<import>` in the error message or somehow heal the situation. But I disagree that a `NotFound` is necessarily better. I think it could be more puzzling than the cyclic error.
> 
> Generally, I think trying `export` for package management is doomed. Even if it worked, I would find it dubious since it would just add another layer of complexity.

I've just run into an error likely caused by this -
What would you consider the use of export other than package management?